lib/host_remove_duplicates.py

```diff
@@ -0,0 +1,166 @@
+def find_host_by_regular_canonical_facts(canonical_facts, query, logger):
+    """
+    Returns all matches for a host containing given canonical facts
+    """
+    logger.debug("find_host_by_regular_canonical_facts(%s)", canonical_facts)
+
+    hosts = multiple_canonical_facts_host_query(canonical_facts, query).order_by(Host.modified_on.desc()).all()
+
+    if hosts:
+        logger.debug("Found existing host using canonical_fact match: %s", hosts)
+
+    return hosts
+
+
+def get_elevated_canonical_facts(canonical_facts):
+    elevated_facts = {
+        key: canonical_facts[key] for key in ELEVATED_CANONICAL_FACT_FIELDS if key in canonical_facts.keys()
+    }
+    return elevated_facts
+
+
+def get_regular_canonical_facts(canonical_facts):
+    regular_cfs = {key: canonical_facts[key] for key in CANONICAL_FACTS if key in canonical_facts.keys()}
+    return regular_cfs
+
+
+def _delete_host(query, host):
+    delete_query = query.filter(Host.id == str(host["id"]))
+    delete_query.delete(synchronize_session="fetch")
+    delete_query.session.commit()
+
+
+def delete_duplicate_hosts(select_query, chunk_size, logger, interrupt=lambda: False):
+    query = select_query
+    logger.info(f"Total number of hosts in inventory: {query.count()}")
+
+    distinct_accounts_query = query.distinct(Host.account)
+    logger.info(f"Total number of accounts in inventory: {distinct_accounts_query.count()}")
+
+    accounts = distinct_accounts_query.limit(chunk_size).all()
+    distinct_accounts = []
+    for acct in accounts:
+        distinct_accounts.append(acct.account)
+
+    for account in distinct_accounts:
+        # set uniquess within the account
+        unique_list = []
+
+        def unique(host):
+            unique_host = {"id": host.id, "account": host.account}
+            if unique_host not in unique_list:
+                unique_list.append(unique_host)
+
+        acct_query = query.filter(Host.account == account)
+        host_list = acct_query.limit(chunk_size).all()
+        for host in host_list:
+            logger.info(f"Host ID: {host.id}")
+            logger.info(f"Canonical facts: {host.canonical_facts}")
+            elevated_cfs = get_elevated_canonical_facts(host.canonical_facts)
+            logger.info(f"elevated canonical facts: {elevated_cfs}")
+            if elevated_cfs:
+                hosts = find_host_by_elevated_canonical_facts(elevated_cfs, acct_query, logger)
+            else:
+                regular_cfs = get_regular_canonical_facts(host.canonical_facts)
+                logger.info(f"regular canonical facts: {regular_cfs}")
+                if regular_cfs:
+                    hosts = find_host_by_regular_canonical_facts(regular_cfs, acct_query, logger)
+
+            unique(hosts[0])
+            logger.info(f"Unique hosts count: {len(unique_list)}")
+            logger.info(f"All hosts count: {len(host_list)}")
+
+        duplicate_list = []
+        for host in host_list:
```

Comment:
This can be done with a set operation. Just make sure the superset is on the left.
```suggestion
        duplicate_list = list(set(host_list) - set(unique_list))
```

Reply:
Just realized you transform the elements of host_list before comparison. You could maybe do that with a map, but at that point I think it's just a little too much reworking, and too little simplifying. Especially for something that already works.